### PR TITLE
Objectcaching

### PR DESCRIPTION
**Pull request by asturur**
- [X] basic caching system
- [x] adapt stateful system to work on different sets of properties
- [x] working cache for text
- [x] caching for groups
- [ ] evaluate performance of caching images  --> not doing it now
- [x] evalualte benefits of having 2 kinds of caching, one involving scaling and one not
- [ ] write tests for changed statefull functions

### CODE REVIEW

**Comment by asturur** [comment]
Cannot run Unit Tests under node with caching on, something to do with creation of additional canvases with jsdom. I do not think we need cache under node, if we are not interacting with objects with mouse or rendering animations, cache is generally slower. So for now is disabled under node. I will write specific tests anyway just to be sure i can track changes.

File: dist/fabric.js [hunk 1]
```diff
@@ -6087,7 +6087,7 @@ fabric.Pattern = fabric.util.createClass(/** @lends fabric.Pattern.prototype */
      * @type Boolean
      * @default
      */
-    stateful: true,
+    stateful: false,
```
Comment (asturur) [inline dist/fabric.js]: default stateful to false. stateful was used in automatic way to detect a change in object modified at end of transformation. Now we detect it properly with the result of actions. Stateful is no more necessary if not for user needs that will activate it. ( detecting changes in fill, stroke and so on)

**Comment by samael205** [comment]
which it is the basic essence of this functionality?

**Comment by asturur** [comment]
having complex object cached on additional canvases. it won t help for rectangles, circles and simple paths, but will help fps/general smoothness in situation where you want to draw lot of complex groups, or styled text.

**Comment by samael205** [comment]
additional canvases?, Hidden canvases? o auxiliar canvases

**Comment by samael205** [comment]
For example, for complex brushes like spray, that render thousands of points to draw the effect?

**Comment by asturur** [comment]
did not try with brushes yet.

**Comment by samael205** [comment]
Ok, when you release, i will try.

**Comment by asturur** [comment]
@samael205  i pushing this change in.
It helps a lot with spray brush, under certain conditions.
I m gonna add a page on the website to explain what this do and how, what are the side effects and how a developer should use it.

**Comment by samael205** [comment]
@asturur Thanks for your Hard work. I will try tomorrow.

**Comment by samael205** [comment]
@asturur, I'm going to work on a totally different project in the probably  threee next months.
I will continue reading the news so I do not miss anything.
See you in a few months

**Pull request merged**